### ISSUE TITLE
Potential Documentation Bug Regarding Type Promotion with Zero-Dimension Tensor

### ISSUE THREAD

**Issue by yewentao256** [pytorch/pytorch#115907]
### 📚 The doc issue

According to the [current documentation](https://pytorch.org/docs/main/tensor_attributes.html#torch-dtype), "If a zero-dimension tensor operand has a higher category than dimensioned operands, we promote to a type with sufficient size and category to hold all zero-dim tensor operands of that category."

However, in practice, when performing an operation like `(int_tensor + long_zerodim)`, where `int_tensor` is `torch.int32` and `long_zerodim` is `torch.int64`, the result unexpectedly remains `torch.int32` instead of being promoted to `torch.int64` as per the documentation.

This behavior suggests a potential inconsistency in either the implementation or the documentation. Could you please clarify this?

**Comment by yewentao256** [comment]
Oh I've understood, the category here means `integral`, so it will not have a dtype promotion.

However, will this cause an overflow when `long_zerodim` is big?

```py
import torch

int_tensor = torch.tensor([1], dtype=torch.int32)
long_zerodim = torch.tensor(2**31, dtype=torch.int64) 

result = int_tensor + long_zerodim
result_value = result.item()
result_dtype = result.dtype

print(result_value, result_dtype)   # -2147483647 torch.int32
```

**Comment by hexronuspi** [comment]
PyTorch does not automatically handle integer overflow. It's the user's responsibility to ensure that operations on integer tensors do not result in overflow. This can be done by choosing an appropriate data type for the tensors, or by checking for potential overflow before performing operations.

```python
import torch

int_tensor = torch.tensor([1], dtype=torch.int32)
long_zerodim = torch.tensor(2**31, dtype=torch.int64) 

result = int_tensor.to(torch.int64) + long_zerodim
result_value = result.item()
result_dtype = result.dtype

print(result_value, result_dtype) #2147483649 torch.int64
```
or, take both of them as 1D vector

```python
import torch

int_tensor = torch.tensor([1], dtype=torch.int32)
long_zerodim = torch.tensor([2**31], dtype=torch.int64) 

result = int_tensor + long_zerodim
result_value = result.item()
result_dtype = result.dtype

print(result_value, result_dtype) #2147483649 torch.int64
```

The value changes as,

```python
int_tensor = torch.tensor([0], dtype=torch.int32)
long_zerodim = torch.tensor([2**31], dtype=torch.int64) 
#-2147483648 torch.int32
```
```python
int_tensor = torch.tensor([0], dtype=torch.int32)
long_zerodim = torch.tensor([2**31], dtype=torch.int64) 
#-2147483647 torch.int32
```
```python
int_tensor = torch.tensor([2**31-1], dtype=torch.int32)
long_zerodim = torch.tensor(2**31, dtype=torch.int64) 
#-1 torch.int32
```

but, 
```python
int_tensor = torch.tensor([-5], dtype=torch.int32)
long_zerodim = torch.tensor(2**31, dtype=torch.int64) 
#2147483643 torch.int32 torch.int32
```

The `negative` sign tells the next step that an overflow might have occoured. When the sum exceeds the maximum value that can be represented by `torch.int32` (2**31 - 1), an overflow occurs, and the result wraps around to the negative side of the torch.int32 range.

**Comment by yewentao256** [comment]
Got it! Thanks a lot! @hexronuspi
